x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/DataStreamIndexSettingsProvider.java

```diff
@@ -21,30 +24,46 @@
 
 public class DataStreamIndexSettingsProvider implements IndexSettingProvider {
 
-    private static final DateFormatter FORMATTER = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER;
+    static final DateFormatter FORMATTER = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER;
 
     @Override
     public Settings getAdditionalIndexSettings(
         String indexName,
         String dataStreamName,
-        boolean newDataStream,
+        Metadata metadata,
         long resolvedAt,
         Settings templateAndRequestSettings
     ) {
-        if (dataStreamName != null && newDataStream) {
+        if (dataStreamName != null) {
             IndexMode indexMode = Optional.ofNullable(templateAndRequestSettings.get(IndexSettings.MODE.getKey()))
                 .map(value -> IndexMode.valueOf(value.toUpperCase(Locale.ROOT)))
                 .orElse(IndexMode.STANDARD);
-            TimeValue lookAheadTime = templateAndRequestSettings.getAsTime(
-                IndexSettings.LOOK_AHEAD_TIME.getKey(),
-                IndexSettings.LOOK_AHEAD_TIME.getDefault(templateAndRequestSettings)
-            );
             if (indexMode == IndexMode.TIME_SERIES) {
-                Instant start = Instant.ofEpochMilli(resolvedAt).minusMillis(lookAheadTime.getMillis());
-                Instant end = Instant.ofEpochMilli(resolvedAt).plusMillis(lookAheadTime.getMillis());
-
+                TimeValue lookAheadTime = templateAndRequestSettings.getAsTime(
+                    IndexSettings.LOOK_AHEAD_TIME.getKey(),
+                    IndexSettings.LOOK_AHEAD_TIME.getDefault(templateAndRequestSettings)
+                );
                 Settings.Builder builder = Settings.builder();
+                DataStream dataStream = metadata.dataStreams().get(dataStreamName);
+                Instant start;
+                if (dataStream == null) {
+                    start = Instant.ofEpochMilli(resolvedAt).minusMillis(lookAheadTime.getMillis());
+                } else {
+                    IndexMetadata currentLatestBackingIndex = metadata.index(dataStream.getWriteIndex());
+                    if (currentLatestBackingIndex.getSettings().hasValue(IndexSettings.TIME_SERIES_END_TIME.getKey()) == false) {
+                        throw new IllegalStateException(
```

Comment:
Doesn't this mean that you cannot switch a data stream from regular mode (such as before TSDB is released) to the timeseries mode, or else rollover won't work due to this exception? We'll likely need a way to allow switching a data stream into this mode without having to recreated the entire data stream, right?

Is this something we'll implement or relax in later work, or should we handle it here?

Reply:
I think we will need to relax it, but I would be ok with doing this as a follow up.

Reply:
No, this error doesn't mean we can't switch a regular data stream to a tsdb data stream. 
I think this check should remain as at least the most recent backing index should have an end_time index setting. At least for now I think this should be the case for migrated data streams, if we learn that we should be lenient here then we can drop this check.

I think for migrating to a tsdb data stream we should have a migrate path (maybe by building on top of migrate or modify data stream APIs), so that we get a data stream in good enough state so that it can be a tsdb data stream.